src/verifier/verifier.rs

```diff
@@ -19,7 +27,535 @@ pub fn build_symbol_table(ast: &mut Whamm, err: &mut ErrorGen) -> SymbolTable {
+                        } else {
+                            self.err.type_check_error(
+                                false,
+                                "Can't get type of argument".to_owned(),
+                                &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                    .map(|l| l.line_col),
+                            );
+                            return None;
+                        }
+                    }
+                } // else function has no arguments
+
+                let fn_name = match *fn_target.clone() {
+                    Expr::VarId { name, .. } => name,
+                    _ => {
+                        self.err.type_check_error(
+                            false,
+                            "Function target must be a variable".to_owned(),
+                            &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                .map(|l| l.line_col),
+                        );
+                        return None;
+                    }
+                };
+
+                if let Some(id) = self.table.lookup(&fn_name) {
+                    if let Some(Record::Fn { .. }) = self.table.get_record(id) {
+                        // TODO how to get the real Datatype of the Param
+                        todo!();
+                    }
+                }
+
+                None
+            }
+            Expr::Ternary {
+                cond,
+                conseq,
+                alt,
+                loc,
+            } => {
+                let cond_ty = self.visit_expr(cond);
+                if let Some(ty) = cond_ty {
+                    if ty != DataType::Boolean {
+                        self.err.type_check_error(
+                            false,
+                            "Condition must be of type boolean".to_owned(),
+                            &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                .map(|l| l.line_col),
+                        );
+                    }
+                }
+                // TODO: should conseq and alt have the same type?
+                let _conseq_ty = self.visit_expr(conseq);
+                let _alt_ty = self.visit_expr(alt);
+                todo!()
+            }
+        }
+    }
+
+    #[allow(unused_variables)]
+    fn visit_datatype(&mut self, datatype: &mut DataType) -> Option<DataType> {
+        unimplemented!()
+    }
+
+    #[allow(unused_variables)]
+    fn visit_value(&mut self, val: &mut Value) -> Option<DataType> {
+        match val {
+            Value::Integer { .. } => Some(DataType::I32),
+            Value::Str { .. } => Some(DataType::Str),
+            Value::Boolean { .. } => Some(DataType::Boolean),
+            Value::Tuple { ty, vals } => {
+                // Alex TODO: this ty does not contain the DataType
+                // and I need to recurse to get the type, why?
```

Comment:
Should be `DataType::Tuple::ty_info`, right?

Reply:
Yes, but it appears to be `Tuple type: Tuple { ty_info: None }`

Reply:
Because the Whamm parser doesn't give the `ty_info` https://github.com/ejrgilbert/whamm/blob/master/src/parser/whamm_parser.rs#L713